oscal-rest-service-app/src/main/java/com/easydynamics/oscalrestservice/api/SspController.java

```diff
@@ -236,18 +254,464 @@ public ResponseEntity<StreamingResponseBody> updateImplementedRequirementPost(
+          .filter(byComp -> incomingByComp.getUuid().equals(byComp.getUuid()))
+          .findAny()
+          .orElse(null);
+    }
+    if (existingStmt == null) {
+      addComponentToList(existingStmt, incomingByComp);
+    } else {
+      try {
+        OscalDeepCopyUtils.deepCopyProperties(existingByComp, incomingByComp);
+      } catch (IllegalAccessException | InvocationTargetException e) {
+        throw new InvalidDataAccessResourceUsageException(
+            "could not deep copy object", e);
+      }
+    }
+
+    logger.debug("SSP Component updated, saving via service");
+    oscalObjectService.save(existingSsp);
+    return makeObjectResponse(incomingByComp, oscalSspByComponentMarshaller);
+  }
+
+  /**
+   * Similar to unmarshallAndValidateId, checks that the given by-component id
+   * matches the UUID in the given json.
+   *
+   * @param id   the request path id
+   * @param json the request body json
+   * @return the unmarshalled object
+   * @throws OscalObjectConflictException when the path ID does not match the body
+   *                                      ID
+   */
+  protected ByComponent oscalSspByComponentMarshaller(String id, String json) {
+    ByComponent incomingOscalObject = oscalSspByComponentMarshaller.toObject(
+        new ByteArrayInputStream(json.getBytes()));
+
+    UUID incomingUuid = incomingOscalObject.getUuid();
+    if (incomingUuid != null && !id.equals(incomingUuid.toString())) {
+      throw new OscalObjectConflictException(incomingUuid.toString(), id);
+    }
+
+    return incomingOscalObject;
+  }
+
+  /**
+   * Does the work of finding an existing SSP and adding
+   * a statements' by-component Id.
+   *
+   * @param id                       the SSP UUID
+   * @param implementedRequirementId the Implemented Requirement uuid
+   * @param statementId              the Statement uuid
+   * @param json                     the SSP contents
+   */
+  private ResponseEntity<StreamingResponseBody> addComponent(
+      String id,
+      String implementedRequirementId,
+      String statementId,
+      String json) {
+    SystemSecurityPlan existingSsp = oscalObjectService.findById(id)
+        .orElseThrow(() -> new OscalObjectNotFoundException(id));
+
+    List<ImplementedRequirement> implReqs =
+        existingSsp.getControlImplementation().getImplementedRequirements();
+    ImplementedRequirement potentialImplReq = null;
+    for (ImplementedRequirement implReq : implReqs) {
+      if (implementedRequirementId.equals(implReq.getUuid().toString())) {
+        potentialImplReq = implReq;
+      }
+    }
+    ImplementedRequirement existingImplReq = potentialImplReq;
+
+    List<Statement> stmts = existingImplReq.getStatements();
+    Statement potentialStmt = null;
+    for (Statement stmt : stmts) {
+      if (statementId.equals(stmt.getUuid().toString())) {
+        potentialStmt = stmt;
+      }
+    }
+    Statement existingStmt = potentialStmt;
```

Comment:
Why do we alias this and then never use the original value?


```suggestion
    Optional<Statement> existingStmt = existingImplReq.getStatements()
      .stream()
      .filter(stmt -> statementId.equals(stmt.getUuid().toString())
      .findAny();
```

Reply:
Originally I wasn't aliasing and just had `exisitingStmt`, but was getting a "defined in an enclosing scope must be final or effectively final" error later. Aliasing avoided this error.  `streams()` and `filters()` are a much better approach.

Reply:
By using `Optional`, methods like `getUUid()` and `getStatement()` cannot be used with this type. Specifically I get the error: "The method getUuid() is undefined for the type Optional<ImplementedRequirement>"

Reply:
How's this instead?

```suggestion
   ImplementedRequirement existingImplReq =
        existingSsp.getControlImplementation().getImplementedRequirements().stream()
        .filter(implReq -> implementedRequirementId.equals(implReq.getUuid().toString()))
        .findAny().orElse(null);
```